### PR TITLE
Replace all semantics for SMT-LIB

### PR DESCRIPTION
**Pull request by Adda0**
This PR fixes the created replace (regex) all operations to correctly handle matching empty strings according to the [SMT-LIB format definition](https://smt-lib.org/theories-UnicodeStrings.shtml).

### CODE REVIEW

**Review by jurajsic — APPROVED** [review]

File: src/nft/strings.cc [hunk 1]
```diff
@@ -487,6 +488,22 @@ Nft nft::strings::replace_reluctant_single_symbol(Symbol from_symbol, Symbol rep
...
+    // ; the result is to prepend t to s.
+    // (str.replace_re String RegLan String String) 
+    //
+    // ; (str.replace_re_all s r t) is the string obtained by replacing,
+    // ; left-to right, each shortest *non-empty* match of r in s by t.
+    // (str.replace_re_all String RegLan String String)
+    if (replace_mode == ReplaceMode::All) {
+        // Removing the empty string from the regex.
+        regex.unify_initial();
+        regex.final.erase(*regex.initial.begin());
```
Comment (jurajsic) [inline src/nft/strings.cc]: This does not work correctly, `unify_initial` may not create a new initial state if `regex` already has only one state.  So for example if we have automaton which contains one state (both initial and final) with a self loop, this will create NFA without final states.

Comment (jurajsic) [reply src/nft/strings.cc]: So either we give `unify_initial()` some special argument that will tell it whether to always force a new initial state or copy the code from `unify_intitial()` here.

Comment (Adda0) [reply src/nft/strings.cc]: Thanks for testing. I will add an argument to the `unify_{initial,final}()` functions.

File: src/nft/strings.cc [hunk 2]
```diff
@@ -487,6 +488,21 @@ Nft nft::strings::replace_reluctant_single_symbol(Symbol from_symbol, Symbol rep
...
+    // ; Note that if the language of r contains the empty string, 
+    // ; the result is to prepend t to s.
+    // (str.replace_re String RegLan String String) 
+    //
+    // ; (str.replace_re_all s r t) is the string obtained by replacing,
+    // ; left-to right, each shortest *non-empty* match of r in s by t.
+    // (str.replace_re_all String RegLan String String)
+    if (replace_mode == ReplaceMode::All) {
+        // Removing the empty string from the regex.
+        regex.unify_initial(true);
```
Comment (jurajsic) [inline src/nft/strings.cc]: You still need to remove the new initial state from the set of final states here.

Comment (Adda0) [reply src/nft/strings.cc]: It was WIP. Fixed in the latest commit with new tests covering these edge cases.

Comment (jurajsic) [reply src/nft/strings.cc]: Ok, let me know when it is finished, so I can look at it again.

File: include/mata/nfa/nfa.hh [hunk 3]
```diff
@@ -138,12 +138,12 @@ public:
...
      * Unify initial states into a single new initial state.
      */
-    void unify_initial();
+    Nfa& unify_initial(bool force_new_state = false);
 
     /**
      * Unify final states into a single new final state.
      */
-    void unify_final();
+    Nfa& unify_final(bool force_new_state = false);
```
Comment (jurajsic) [inline include/mata/nfa/nfa.hh]: I would update the comments (the new argument + returns `this`).

Comment (Adda0) [reply include/mata/nfa/nfa.hh]: Good point. Fixed.

**Comment by jurajsic** [comment]
I think you can merge.
